### ISSUE TITLE
Posible deadlock with Java 21 Virtual Threads

### ISSUE THREAD

**Issue by Moscagus** [open-telemetry/opentelemetry-java-instrumentation#10747]
Java 21
Opentelemetry Java Agent 2.1.0
Opentelemetry SDK 1.35.0

I have an application in java21 that makes heavy use of virtual threads. After the stress test, he ended up stuck. After an investigation I verified that the problem came from the java opentelemetry agent, even in its latest version 2.1.0. The problem arises because the agent leaves the carrier thread in an inconsistent state.

[jcmd-otel.txt](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/14470332/jcmd-otel.txt)
[jstack-otel.txt](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/14470333/jstack-otel.txt)

**1 - "jstack-otel.txt" show the problem. For example:**
...
"ForkJoinPool-1-worker-1" **#134** [271] daemon prio=5 os_prio=0 cpu=70474.35ms elapsed=722.35s allocated=14189M defined_classes=272 tid=0x000055d55a76ef20  [0x00007f75af255000]
   Carrying virtual thread **#134** --> **Bug: same TID as carrier**
..

**2 - "jcmd-otel.txt" show the stack with problem. For example:**
..
           "tid": "159185",
           "name": "virtual-Tn3EventsConsumerService-158213",
           "stack": [
              **"java.base\/**jdk.internal.misc.Unsafe.park**(Native Method)",   --> **Bug: VT use  java.lang.VirtualThread.park, no jdk.internal.misc.Unsafe.park**
              "java.base\/java.util.concurrent.locks.LockSupport.park(LockSupport.java:221)",
              "java.base\/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:754)",
              "java.base\/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:990)",
              "java.base\/java.util.concurrent.locks.ReentrantLock$Sync.lock(ReentrantLock.java:153)",
              "java.base\/java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:322)",
              **"java.base\/java.lang.ref.ReferenceQueue.poll(ReferenceQueue.java:182)",**
              ****"**io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.AbstractWeakConcurrentMap.expungeStaleEntries**(AbstractWeakConcurrentMap.java:235)",**  --> opentelemetry java agent**
              "io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.WeakConcurrentMap$WithInlinedExpunction.getIfPresent(WeakConcurrentMap.java:193)",
              "io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.WeakLockFreeCache.get(WeakLockFreeCache.java:26)",
              "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.mapGet(VirtualFieldImplementationsGenerator.java:298)",
              "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.realGet(VirtualFieldImplementationsGenerator.java)",
              "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.get(VirtualFieldImplementationsGenerator.java:280)",
              "io.opentelemetry.javaagent.bootstrap.executors.ExecutorAdviceHelper.cleanPropagatedContext(ExecutorAdviceHelper.java:92)",
              "java.base\/java.util.concurrent.FutureTask.cancel(FutureTask.java:181)",
              "java.base\/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.cancel(ScheduledThreadPoolExecutor.java:291)",
              **"java.base\/**java.lang.VirtualThread.cancel**(VirtualThread.java:705)",**
              **"java.base\/**java.lang.VirtualThread.parkNanos**(VirtualThread.java:628)",** 
...

**3 - **VirtualThread.java:**** https://github.com/openjdk/loom/blob/fibers/src/java.base/share/classes/java/lang/VirtualThread.java#L681
https://github.com/openjdk/loom/blob/fibers/src/java.base/share/classes/java/lang/VirtualThread.java#L767

    @Override
    void parkNanos(long nanos) {
        ...
        if (nanos > 0) {
            ...
            try {
                yielded = yieldContinuation();  // may throw
            } finally {
                ...
                cancel(unparker);
            }
            ...
            }
        }
    }

    @ChangesCurrentThread
    private void cancel(Future<?> future) {
        ...
        if (!future.isDone()) {                --------> When this condition occurs, the problem is generated
            // need to switch to current carrier thread to avoid nested parking
            switchToCarrierThread();     --------> Here the carrier TID and the carrying TID are the same as shown in point 1 (jstack-otel.txt)
            try {
                future.cancel(false);        --------> Here the opentelemetry java agent begins to act (Bug).
            } finally {
                switchToVirtualThread(this);  ---------> It never comes here
            }
        }
    }

**4 - AbstractWeakConcurrentMap.java**
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/release/v2.1.x/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/cache/weaklockfree/AbstractWeakConcurrentMap.java#L232

    public static void expungeStaleEntries() {
      Reference<?> reference;
      while ((reference = REFERENCE_QUEUE.poll()) != null) {
         removeWeakKey((WeakKey<?>) reference);
      }
    }

**REFERENCE_QUEUE.poll()**   -------> poll() that generates **jdk.internal.misc.Unsafe.park**, thus achieving a stuck in the application since as explained in point 1:
"ForkJoinPool-1-worker-1" **#134** [271] daemon prio=5 os_prio=0 cpu=70474.35ms elapsed=722.35s allocated=14189M defined_classes=272 tid=0x000055d55a76ef20  [0x00007f75af255000]
   Carrying virtual thread **#134** --> **Bug: same TID as carrier**

**Comment by JonasKunz** [comment]
Hey @Moscagus, thanks for the detailed analysis!
To my understanding this issue is actually slightly different from #10181 because here only certain threads hang instead of the entire JVM, correct?

The root cause of this issue that we try to acquire a contended lock while a thread is about to mount or unmount.
When such a lock is contended, the JVM will try to unmount the current thread while it is already in the process of being mounted / unmounted. This causes the thread to basically deadlock with itself.

I've experienced the same issue while playing around with JVMTI mount / unmount callbacks.

Note that this issue doesn't occur (yet) with `synchronized` locking, because those currently do not cause virtual threads to be unmounted. However, to my knowledge the JVM devs are working on lifting this restriction, so we'll see what the future holds here.

The deadlock you are seeing here is as you already analyzed because the virtual thread implementation internally seems to use a standard `ScheduledThreadPool` for scheduling virtual threads, 
see https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/VirtualThread.java#L1237.

This thread-pool is instrumented by the opentelemetry agent for context propagation.

IMO the best solution would be to not instrument those internal pools. That is however not really easily doable to my knowledge, because it uses standard classes which we can't just exclude.
We would need to build some exclusion on instance level.

Another safe option would be to make the instrumentation No-Op when it runs in a Virtual Thread mounting / unmounting context.
To my knowledge this would require instrumenting very low-level and frequently called methods, such as `VirtualThread.parkNanos` or walking the stack, which would be pretty expensive.
In addition this would make us highly dependent on the internal loom implementation.

A simple workaround for now would be to move the cleanup of `VirtualField` `WeakConcurrentHashMap`'s stale entries to a separate thread which performs a periodic cleanup.
[Here](https://github.com/elastic/elastic-otel-java/blob/main/common/src/main/java/co/elastic/otel/common/WeakConcurrent.java#L33) is an example implementation how to this.
This has the side benefit of reducing the latency of get/put calls performed on application threads. For this reason the elastic-apm-agent has already been using a separate cleaner thread.

This will prevent the issue from occurring because the `ReferenceQueue` won't be polled anymore during mount/unmount of a virtual thread.
The underlying `WeakHashMap` might still be contented, it however to my knowledge uses `synchronized` instead of other locks, making this a non-issue.

Another way of circumvent this would be to use wrapping of `Runnables` / `Futures` instead of `VirtualField`s, for which already an issue exists: #9324.

I'd like some more Feedback here on which route to take. Based on that I should be able to find the time to implement a fix.

**Comment by Moscagus** [comment]
Hi @JonasKunz, I think there are 2 issues:

1 - **expungeStaleEntries**:  I agree with you that a good workaround  would be to move the cleanup of VirtualField WeakConcurrentHashMap's stale entries to a separate thread which performs a periodic cleanup.

2 - **getIfPresent**:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/release/v2.1.x/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/cache/weaklockfree/WeakConcurrentMap.java#L191

    @Override
    public V getIfPresent(K key) {
      expungeStaleEntries();    ------> It would be moved to a separate thread which performs a periodic cleanup.
      return super.getIfPresent(key);    ------->  AbstractWeakConcurrentMap.java
    }

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/release/v2.1.x/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/cache/weaklockfree/AbstractWeakConcurrentMap.java#L111

    public V getIfPresent(K key) {
      if (key == null) {
        throw new NullPointerException();
      }
      L lookupKey = getLookupKey(key);    ----> WeakConcurrentMap.java
      try {
        return target.get(lookupKey);
      } finally {
        resetLookupKey(lookupKey);
      }
    }

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/release/v2.1.x/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/cache/weaklockfree/WeakConcurrentMap.java#L119

    @Override
    @SuppressWarnings("unchecked")
    protected LookupKey<K> getLookupKey(K key) {
      LookupKey<K> lookupKey;
      if (reuseKeys) {
        lookupKey = (LookupKey<K>) LOOKUP_KEY_CACHE.get();
      } else {
        lookupKey = new LookupKey<>();
      }
      return lookupKey.withValue(key);
    }

**LOOKUP_KEY_CACHE is ThreadLocal**

    private static final ThreadLocal<LookupKey<?>> LOOKUP_KEY_CACHE =
        new ThreadLocal<LookupKey<?>>() {
          @Override
          protected LookupKey<?> initialValue() {
            return new LookupKey<>();
          }
        };

I may be wrong in my analysis, but I'm wondering if LOOKUP_KEY_CACHE, which should be local to the Virtual Thread, is okay to be consumed by the Carrier Thread ((LOOKUP_KEY_CACHE.get()).
If I'm not mistaken, this is another point to resolve.

**Comment by laurit** [comment]
> A simple workaround for now would be to move the cleanup of VirtualField WeakConcurrentHashMap's stale entries to a separate thread which performs a periodic cleanup.

We already have a background thread that cleans these maps https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/cache/weaklockfree/AbstractWeakConcurrentMap.java so removing the inline expunge is an option.

**Comment by JonasKunz** [comment]
> I may be wrong in my analysis, but I'm wondering if LOOKUP_KEY_CACHE, which should be local to the Virtual Thread, is okay to be consumed by the Carrier Thread ((LOOKUP_KEY_CACHE.get()).
If I'm not mistaken, this is another point to resolve.

I think that shouldn't be a a problem here: The way the `LOOKUP_KEY_CACHE` is used, it is impossible for the thread to be unmounted while the thread currently uses the cached `LookupKey` for a lookup. The `WeakHashMap` implementation doesn't care about the values of the cached `LookupKey` before or after this usage.

> so removing the inline expunge is an option.

I'll try to do that and open a PR (soon hopefully). I'll also try if I can reproduce the deadlock in a unit test first, so that we can more easily detect problems with future iterations of Project Loom here.

**Comment by JonasKunz** [comment]
@Moscagus Unfortunately I haven't been able to reproduce the issue.
Could you try the [snapshot](https://github.com/open-telemetry/opentelemetry-java-instrumentation/actions/runs/8171379103/artifacts/1301934207) from #10768 and verify that it fixes the issue?

**Comment by Moscagus** [comment]
Unfortunately it's still using "expungeStaleEntries"
I decompile the jar and I actually saw the changes, but when it runs it still continues creating "WeakConcurrentMap$WithInlinedExpunction"

[jcmd-otel.txt](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/14529242/jcmd-otel.txt)
[jstack-otel.txt](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/14529243/jstack-otel.txt)

I keep investigating

**Comment by Moscagus** [comment]
I think the problem is in:

    private static final class CacheBasedVirtualFieldSupplier implements VirtualFieldSupplier {

      private final Cache<Class<?>, Cache<Class<?>, VirtualField<?, ?>>>
          ownerToFieldToImplementationMap = Cache.weak();    ---> new WeakLockFreeCache<>(true) = this.delegate = (WeakConcurrentMap<K, V>)new WeakConcurrentMap.WithInlinedExpunction();

      @Override
      @SuppressWarnings("unchecked")
      public <U extends T, V extends F, T, F> VirtualField<U, V> find(
          Class<T> type, Class<F> fieldType) {
        return (VirtualField<U, V>)
            ownerToFieldToImplementationMap
                .computeIfAbsent(type, c -> Cache.weak())
                .computeIfAbsent(fieldType, c -> new CacheBasedVirtualField<>());
      }
  }


Therefore, every time you call **find()**:

.computeIfAbsent(fieldType, c -> new CacheBasedVirtualField<>());

**computeIfAbsent in WeakConcurrentMap.WithInlinedExpunction**=
      public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
        expungeStaleEntries();   ---> POLL again
        return (V)super.computeIfAbsent(key, mappingFunction);
      }

**Comment by Moscagus** [comment]
On the other hand, the finds are regenerated by Byte Buddy, and from what I see they will always use "Cache.weak()"

       "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.mapGet(VirtualFieldImplementationsGenerator.java:298)",
              "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.realGet(VirtualFieldImplementationsGenerator.java)",
              "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.get(VirtualFieldImplementationsGenerator.java:280)",

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/field/VirtualFieldFindRewriter.java#L42

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/field/VirtualFieldImplementationsGenerator.java#L269

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/field/VirtualFieldImplementationsGenerator.java#L312

**Comment by JonasKunz** [comment]
I missed out on the fact that `RuntimeVirtualFieldSupplier.CacheBasedVirtualField` isn't used for Advice `VirtualField.find` calls, even if it was not possible for the JVM to inject fields.

I've fixed this now in my PR, @Moscagus could you try again with the [new snapshot](https://github.com/open-telemetry/opentelemetry-java-instrumentation/actions/runs/8250468313/artifacts/1318916477) ?

**Comment by JonasKunz** [comment]
Update after a discussion with @laurit : My fix likely won't actually solve the problem, even if it doesn't surface anymore.
`ConcurrentHashMap` doesn't rely on `synchronized` only, but also calls `LockSupport.park` in contended cases.
That would result in the same nested unparking attempts under contention like polling the `ReferenceQueue`.

**Comment by Moscagus** [comment]
Exactly as you mention the problem is still the same as before:

              "io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.AbstractWeakConcurrentMap.expungeStaleEntries(AbstractWeakConcurrentMap.java:235)",
              "io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.WeakConcurrentMap$WithInlinedExpunction.getIfPresent(WeakConcurrentMap.java:193)",
              "io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.WeakLockFreeCache.get(WeakLockFreeCache.java:30)",
              "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.mapGet(VirtualFieldImplementationsGenerator.java:298)",
              "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.realGet(VirtualFieldImplementationsGenerator.java)",
              "io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$java$util$concurrent$Future$io$opentelemetry$javaagent$bootstrap$executors$PropagatedContext.get(VirtualFieldImplementationsGenerator.java:280)",
              "io.opentelemetry.javaagent.bootstrap.executors.ExecutorAdviceHelper.cleanPropagatedContext(ExecutorAdviceHelper.java:92)",

**Comment by laurit** [comment]
@Moscagus could you try with https://github.com/open-telemetry/opentelemetry-java-instrumentation/actions/runs/8277540924/artifacts/1325242993 an let us know whether this changes anything?

**Comment by Moscagus** [comment]
@laurit after several consecutive tests I confirm that the solution does not work.
[jcmd.txt](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/14616891/jcmd.txt)
[jstack.txt](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/14616892/jstack.txt)

I keep investigating

**Comment by Moscagus** [comment]
Could it be that since "VirtualThread" is final it cannot be instrumented ?

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/VirtualThread.java#L68

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/executors/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/executors/VirtualThreadInstrumentation.java#L21

**Comment by laurit** [comment]
> Could it be that since "VirtualThread" is final it cannot be instrumented ?
> 
> https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/VirtualThread.java#L68
> 
> https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/executors/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/executors/VirtualThreadInstrumentation.java#L21

No, `final` modifier does not prevent instrumenting. I think we need to find a way to reproduce this.

**Comment by Moscagus** [comment]
@laurit what do you think about changing the check to: **!Thread.currentThread().isVirtual()**

    private static boolean isPropagationDisabled() {
      return (!Thread.currentThread().isVirtual());
    }

If this is valid, there would be no need for the "ThreadLocal propagationDisabled" variable and of the "VirtualThreadInstrumentation" class.

**Comment by Moscagus** [comment]
Sorry, I'm thinking as if all applications were on Java 21 using virtual threads only.

I don't think what I just mentioned helps.

**Comment by 123liuziming** [comment]
I have a question about this issue, the platform thread is parked because it acquires the AQS lock, but why is the platform thread not unparked after the other thread releases the AQS lock. It seems that the platform thread always hangs in there.

**Comment by 123liuziming** [comment]
```
/*
 * Copyright (c) 1997, 2022, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.lang.ref;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;
import jdk.internal.misc.VM;

/**
 * Reference queues, to which registered reference objects are appended by the
 * garbage collector after the appropriate reachability changes are detected.
 * @param <T> the type of the reference object
 *
 * @author   Mark Reinhold
 * @since    1.2
 */

public class ReferenceQueue<T> {
    private static class Null extends ReferenceQueue<Object> {
        public Null() { super(0); }

        @Override
        boolean enqueue(Reference<?> r) {
            return false;
        }
    }

    static final ReferenceQueue<Object> NULL = new Null();
    static final ReferenceQueue<Object> ENQUEUED = new Null();

    private volatile Reference<? extends T> head;
    private long queueLength = 0;

    private final ReentrantLock lock;
    private final Condition notEmpty;

    void signal() {
        notEmpty.signalAll();
    }

    void await() throws InterruptedException {
        notEmpty.await();
    }

    void await(long timeoutMillis) throws InterruptedException {
        notEmpty.await(timeoutMillis, TimeUnit.MILLISECONDS);
    }

    /**
     * Constructs a new reference-object queue.
     */
    public ReferenceQueue() {
        this.lock = new ReentrantLock();
        this.notEmpty = lock.newCondition();
    }

    ReferenceQueue(int dummy) {
        this.lock = null;
        this.notEmpty = null;
    }

    final boolean enqueue0(Reference<? extends T> r) { // must hold lock
        // Check that since getting the lock this reference hasn't already been
        // enqueued (and even then removed)
        ReferenceQueue<?> queue = r.queue;
        if ((queue == NULL) || (queue == ENQUEUED)) {
            return false;
        }
        assert queue == this;
        // Self-loop end, so if a FinalReference it remains inactive.
        r.next = (head == null) ? r : head;
        head = r;
        queueLength++;
        // Update r.queue *after* adding to list, to avoid race
        // with concurrent enqueued checks and fast-path poll().
        // Volatiles ensure ordering.
        r.queue = ENQUEUED;
        if (r instanceof FinalReference) {
            VM.addFinalRefCount(1);
        }
        signal();
        return true;
    }

    final boolean headIsNull() {
        return head == null;
    }

    final Reference<? extends T> poll0() { // must hold lock
        Reference<? extends T> r = head;
        if (r != null) {
            r.queue = NULL;
            // Update r.queue *before* removing from list, to avoid
            // race with concurrent enqueued checks and fast-path
            // poll().  Volatiles ensure ordering.
            @SuppressWarnings("unchecked")
            Reference<? extends T> rn = r.next;
            // Handle self-looped next as end of list designator.
            head = (rn == r) ? null : rn;
            // Self-loop next rather than setting to null, so if a
            // FinalReference it remains inactive.
            r.next = r;
            queueLength--;
            if (r instanceof FinalReference) {
                VM.addFinalRefCount(-1);
            }
            return r;
        }
        return null;
    }

    final Reference<? extends T> remove0(long timeout)
            throws IllegalArgumentException, InterruptedException { // must hold lock
        Reference<? extends T> r = poll0();
        if (r != null) return r;
        long start = System.nanoTime();
        for (;;) {
            await(timeout);
            r = poll0();
            if (r != null) return r;

            long end = System.nanoTime();
            timeout -= (end - start) / 1000_000;
            if (timeout <= 0) return null;
            start = end;
        }
    }

    final Reference<? extends T> remove0() throws InterruptedException { // must hold lock
        for (;;) {
            var r = poll0();
            if (r != null) return r;
            await();
        }
    }

    boolean enqueue(Reference<? extends T> r) { /* Called only by Reference class */
        lock.lock();
        try {
            return enqueue0(r);
        } finally {
            lock.unlock();
        }
    }

    /**
     * Polls this queue to see if a reference object is available.  If one is
     * available without further delay then it is removed from the queue and
     * returned.  Otherwise this method immediately returns {@code null}.
     *
     * @return  A reference object, if one was immediately available,
     *          otherwise {@code null}
     */
    public Reference<? extends T> poll() {
        if (headIsNull())
            return null;
        lock.lock();
        try {
            return poll0();
        } finally {
            lock.unlock();
        }
    }

    /**
     * Removes the next reference object in this queue, blocking until either
     * one becomes available or the given timeout period expires.
     *
     * <p> This method does not offer real-time guarantees: It schedules the
     * timeout as if by invoking the {@link Object#wait(long)} method.
     *
     * @param  timeout  If positive, block for up to {@code timeout}
     *                  milliseconds while waiting for a reference to be
     *                  added to this queue.  If zero, block indefinitely.
     *
     * @return  A reference object, if one was available within the specified
     *          timeout period, otherwise {@code null}
     *
     * @throws  IllegalArgumentException
     *          If the value of the timeout argument is negative
     *
     * @throws  InterruptedException
     *          If the timeout wait is interrupted
     */
    public Reference<? extends T> remove(long timeout) throws InterruptedException {
        if (timeout < 0)
            throw new IllegalArgumentException("Negative timeout value");
        if (timeout == 0)
            return remove();

        lock.lock();
        try {
            return remove0(timeout);
        } finally {
            lock.unlock();
        }
    }

    /**
     * Removes the next reference object in this queue, blocking until one
     * becomes available.
     *
     * @return A reference object, blocking until one becomes available
     * @throws  InterruptedException  If the wait is interrupted
     */
    public Reference<? extends T> remove() throws InterruptedException {
        lock.lock();
        try {
            return remove0();
        } finally {
            lock.unlock();
        }
    }

    /**
     * Iterate queue and invoke given action with each Reference.
     * Suitable for diagnostic purposes.
     * WARNING: any use of this method should make sure to not
     * retain the referents of iterated references (in case of
     * FinalReference(s)) so that their life is not prolonged more
     * than necessary.
     */
    void forEach(Consumer<? super Reference<? extends T>> action) {
        for (Reference<? extends T> r = head; r != null;) {
            action.accept(r);
            @SuppressWarnings("unchecked")
            Reference<? extends T> rn = r.next;
            if (rn == r) {
                if (r.queue == ENQUEUED) {
                    // still enqueued -> we reached end of chain
                    r = null;
                } else {
                    // already dequeued: r.queue == NULL; ->
                    // restart from head when overtaken by queue poller(s)
                    r = head;
                }
            } else {
                // next in chain
                r = rn;
            }
        }
    }
}

```
This is the `ReferenceQueue` in JDK21, it seems that the lock is always released in `finally` block. So why the virtual thread blocked on the lock?

**Comment by trask** [comment]
> It seems that the platform thread always hangs in there.

hi @123liuziming, can you open a new issue for this and include a repro that folks can use to investigate? thanks

**Comment by 123liuziming** [comment]
> > It seems that the platform thread always hangs in there.
> 
> hi @123liuziming, can you open a new issue for this and include a repro that folks can use to investigate? thanks

I 've figured it out. Thanks!

**Comment by YaoYingLong** [comment]
> [@Moscagus](https://github.com/Moscagus) Unfortunately I haven't been able to reproduce the issue. Could you try the [snapshot](https://github.com/open-telemetry/opentelemetry-java-instrumentation/actions/runs/8171379103/artifacts/1301934207) from [#10768](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10768) and verify that it fixes the issue?

@JonasKunz Have you already reproduced this issue? Is there a demo example for reproducing the issue? I encountered the same problem, but I haven't been able to reproduce it.

**Comment by JonasKunz** [comment]
No, we weren't able to reproduce the issue and assumed that #10881 was successful in fixing it.
